### PR TITLE
Refactor fields (step 4)

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

```SUMMARY: Infrastructure "Refactor fields (step 4)"```

#### Purpose of change

Unhardcode data structures to allow field type customization/addition via json.

#### Describe the solution

Current pull request does the following:

- translates lambdas into normal functions (no visible changes from player side - just moving some code around).

### CODE REVIEW

File: src/map.h [hunk 1]
```diff
@@ -29,6 +29,7 @@
 #include "cata_utility.h"
 #include "faction.h"
 #include "point.h"
+#include "scent_block.h"
```
Comment (BevapDin) [inline src/map.h]: If you only need this for `scent_block` in `spread_gas`, a forward declaration should be enough.

File: src/map.h [hunk 2]
```diff
@@ -370,6 +371,14 @@ class map
         // Versions of the above that don't do bounds checks
         const maptile maptile_at_internal( const tripoint &p ) const;
         maptile maptile_at_internal( const tripoint &p );
+        const maptile maptile_has_bounds( const tripoint &p, const bool bounds_checked );
+        const std::array<maptile, 8> get_neighbors( const tripoint &p );
```
Comment (BevapDin) [inline src/map.h]: Why does this return a `const` object? (Same for the function above.)

File: src/map_field.cpp [hunk 3]
```diff
@@ -154,97 +171,188 @@ static int ter_furn_movecost( const ter_t &ter, const furn_t &furn )
...
+        if( gas_can_spread_to( cur, up_tile ) && valid_move( p, up, true, true ) ) {
+            gas_spread_to( cur, up_tile );
         }
     }
+};
 
-    point index( const tripoint &p ) const {
-        return point( p.x - origin.x, p.y - origin.y );
+/*
+Function: create_hot_air
```
Comment (BevapDin) [inline src/map_field.cpp]: Why are you repeating the name of the function in the comment?

Comment (ZhilkinSerg) [reply src/map_field.cpp]: I've just moved this code block, but yeah comments should be revised.

**Comment by ZhilkinSerg** [comment]
Jenkins rebuild.

**Pull request merged**